fixes #10467 - parameter override ui fixes

unorthodoxgeek opened the pull request.

unorthodoxgeek committed:
fixes #10467 - parameter override ui fixes

domcleal commented on db/migrate/20150602114114_add_hidden_flag_to_lookup_values.rb:
```diff
@@ -0,0 +1,5 @@
+class AddHiddenFlagToLookupValues < ActiveRecord::Migration
+  def change
+    add_column :lookup_values, :hidden, :boolean, :default => false
```
Shouldn't this be on the lookup key, not the value?  I don't think you'd want the risk of missing it on one override when it could simply be set for the parameter as a whole.

domcleal replied:
Still an issue.

unorthodoxgeek replied:
true. I'll move it.

domcleal commented:
1) The number of navigation levels, titles and tabs on the host edit page is too many, it really needs reducing.

![screenshot from 2015-06-09 14 16 53](https://cloud.githubusercontent.com/assets/482089/8058576/462bba4a-0eb2-11e5-8bb7-5a59675a6f2e.png)

2) The host group edit page looks a bit of a mess on the Parameters tab - it's not on separate tabs, just one big vertical list, with different titles to the host page.

![screenshot from 2015-06-09 14 18 23](https://cloud.githubusercontent.com/assets/482089/8058611/803c8232-0eb2-11e5-8069-4478321f72b9.png)

unorthodoxgeek commented:
@tbrisker - I've messed about with your code for that fullscreen button. Please test this and make sure that whatever you did with it didn't break by my changes. thanks :smile:

Also, why is the JS obtrusive? :cry:

unorthodoxgeek commented:
@domcleal regarding (1) - already discussed it with Kyle and decided that this is the less horrible option, changed the type of tabs so it will be less confusing.
fixed the hostgroup form.

ShimShtein commented:
I saw @domcleal's screenshot of the hostgroup form, the "hidden" and "puppet default" flags look not too well, is it possible to create a single line from the text, and put two push-buttons aligned to it, like this:
![image](https://cloud.githubusercontent.com/assets/10046168/8082566/e1600d66-0f86-11e5-9f87-f81685fb5174.png)

dLobatog commented on app/views/common_parameters/_parameters_table.html.erb:
```diff
@@ -0,0 +1,27 @@
+    </ul>
+    <div class="tab-content">
+      <div class="tab-pane active" id="puppet_parameters">
+        <h4><%= _('Class Parameters') %></h4>
```
Isn't this one redundant? The tab already says "Class parameters"

dLobatog commented on app/views/hostgroups/_form.html.erb:
```diff
@@ -64,20 +64,34 @@
+        </ul>
+        <div class="tab-content">
+          <div class="tab-pane active" id="puppet_parameters">
+            <h4><%= _('Class Parameters') %></h4>
```
Same as in Host, I believe this is redundant information

dLobatog commented on app/views/hosts/_form.html.erb:
```diff
@@ -86,19 +86,7 @@
-        <p></p>
-        <%= render "common_parameters/parameters", { :f => f, :type => :host_parameters } %>
-      </div>
+      <%= render('common_parameters/parameters_table', :f => f, :obj => @host, :inherited_parameters => @host.host_inherited_params(true)) %>
```
:clap:

dLobatog commented:
@unorthodoxgeek Something that I believe has been broken with parameters for a long time is this (which could be solved in this PR or a separate one).

![](http://i.imgur.com/2t85FJc.png) 

Even though we're overriding or adding parameters at the Host level -same goes for Host groups- they are displayed in the table as "Global parameters" which is false, Global parameters can only be created through `/common_parameters`. 

The issue stems from using the same template for both things, but if you could find a way to show these are Host or Host group parameters in this dialog it'd be correct.

domcleal commented:
> @domcleal regarding (1) - already discussed it with Kyle and decided that this is the less horrible option, changed the type of tabs so it will be less confusing.

1) I'd say it's a step backwards from the current implementation then.  Removing that additional header has helped, but it's still a lot of extra levels of navigation, especially if you have a normal number of classes.

Edit: to add to my last point, the new layout adds an extra click for every class assigned to the host.  It also makes it very hard to see an overview with the data spread throughout many tabs.

2) The parameters tab doesn't seem to update any more when you add/remove Puppet classes.

3) If a class has no overridden parameters, it shows an empty list.

![screenshot from 2015-06-11 08 43 06](https://cloud.githubusercontent.com/assets/482089/8102183/efc34c58-1015-11e5-80fb-302b4b8facc2.png)

4) When a host has no classes, a completely blank tab is shown.

![screenshot from 2015-06-11 08 45 08](https://cloud.githubusercontent.com/assets/482089/8102230/41a098d2-1016-11e5-8d9d-a1cb89d4ce6b.png)

5) The two blue tab boxes are touching when the topmost class is selected, which looks odd to me - missing some spacing?

![screenshot from 2015-06-11 08 43 50](https://cloud.githubusercontent.com/assets/482089/8102200/101753dc-1016-11e5-9e30-a3cb4fef6531.png)

tbrisker commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,61 @@
+                            :id => "override_for_#{key.id}",
+                            :placeholder => _("Value")) %>
+              <% end %>
+              <%= fullscreen_button("'#override_for_#{key.id}'")%>
```
Please use  `<%= fullscreen_button("$('#override_for_#{key.id}')")%>` instead of adding `element=$(element)` in the JS function. 
This is done this way because most buttons use `$(this)` to get their relative position in the DOM and traverse from it to the element they target.

unorthodoxgeek replied:
can't. the element changes between textarea and input

tbrisker replied:
So what? it only means the element matching is done by the button instead of in the function. It still matches the element with the given id. (even though it makes no sense having a fullscreen button for a hidden input...)

unorthodoxgeek replied:
nope, doesn't work. $(foo) means that element, if it's replaced it's no longer pointing at anything. tried it already

unorthodoxgeek replied:
fixed. thanks :+1:

tbrisker commented on app/assets/stylesheets/application.scss:
```diff
@@ -18,6 +18,8 @@ body {
   height: 100%;
 }
 
+.clearfix {clear:both;}
```
I think bootstrap already has a `.clearfix` class.

unorthodoxgeek replied:
yes, and somewhere along the way we broke it, and I can't find it.

waldenraines replied:
The `.clearfix` class appears to be working for me in the `develop` branch.

Also, this class isn't being used in this PR.

unorthodoxgeek replied:
without this, the form looks like this: (notice the cancel button is not cleared for some reason)
![screen shot 2015-06-18 at 12 19 51 pm](https://cloud.githubusercontent.com/assets/706359/8228027/5e53feb2-15b4-11e5-982c-6b52dbe75012.png)

waldenraines replied:
I still believe we should find the source of this issue instead of putting in the above.  Unfortunately I don't have any suggestions at the moment.

waldenraines replied:
Also, this button aligns correctly for me, although I'm not sure why everything is empty.

![screen shot 2015-07-07 at 2 03 19 pm](https://cloud.githubusercontent.com/assets/4116405/8553383/152b1ae8-24b1-11e5-8cd3-eb6ce311fdbf.png)

Perhaps the problem lies in the content of that box?  It almost seems like a closing tag is missing somewhere.

waldenraines commented on app/assets/stylesheets/hosts.scss:
```diff
@@ -35,4 +35,40 @@
+    font-weight: normal;
+  }
+  .btn-sm {
+    margin-top: -4px;
```
Is there another way this can be accomplished without the negative margins?

unorthodoxgeek replied:
not that I know of...

waldenraines commented on app/views/hostgroups/_form.html.erb:
```diff
@@ -64,20 +64,31 @@
       <%= password_f f, :root_pass, :help_inline => _("Password must be 8 characters or more")%>
     </div>
 
+
```
Nitpick but any reason for this extra line?

unorthodoxgeek commented:
@domcleal - updated

domcleal commented:
> @domcleal - updated

Points 1, 2, 3 and 4 don't seem to have been addressed, only point 5.  Can you confirm please?

unorthodoxgeek commented:
@domcleal - 3 and 4 are now fixed (when no lookup keys go with a class, it doesn't show up, when no classes are there, the active tab would be the global params), I couldn't reproduce 2 - when a class has lookup keys related to it it shows up as expected for me.
regarding 1 - the design I implemented is what was given to me by Kyle Baker, your concerns were passed on to him, he made the corrections he felt he should make.

orrabin commented:
@domcleal The new layout does add an extra click but it helps when you are editing parameters for a specific puppetclass. Today if you have 4-5 parameters you allow to override in a class and you add 3 classes you will have to scroll up and down to find the original value and the override on the host.
We assumed when presenting to Kyle what we want from the design that we want clearness for each puppetclass - what are the parameters you can override and what is their value for this host.
The previous layout gave you a immediate overview of all the parameters but it was hard to read it as is.

I'm not against changing this if you have a suggestion on how to get both a clear view of a certain puppetclass and an overview. We thought focusing on each puppetclass while editing the host is the important thing hence we preferred this design.
The suggestion to change this to look like mail notifications was a bit overwhelming with 10-20 parameters on the same page.
If our usecase is wrong or you know of different ones we are missing that can help to revise the design.

domcleal commented:
re 4) although the active tab has changed, that's an improvement, I think having the completely blank tab and big empty box is still not very pretty.  I'd suggest removing the tab or adding text in there at least.

> I couldn't reproduce 2 - when a class has lookup keys related to it it shows up as expected for me.

This is still an issue for me, any time I add a class from the Puppet classes tab, I see a /parameters call to the server but no change on the classes tab.  I don't see _any_ call to /parameters when a class is removed.  Note that I had to rebase the branch onto develop to be able to run it.

I see the DB migration also changed to lookup keys, however the lookup key hidden field doesn't seem to be shown on the Puppet class edit page for the parameter, and it's still listed on each override rather than the key/param.

> We thought focusing on each puppetclass while editing the host is the important thing hence we preferred this design.

I'm unsure that this would be universally useful - how it's used is going to vary a lot based on class composition, and this change almost always is going to result in a lot of clicking, which I'm not sure is an improvement either way.

domcleal commented:
6) When saving a host with overrides and saving fails, two bad behaviours: a) the override doesn't remain overridden on the page reload, b) clicking the Override button again redirects the page to /hosts.

7) "Use Puppet default" ought to disable the override value text box as it does in the Puppet class edit.

8) Text boxes for default + override value, and full screen buttons probably ought to be the same size and location horizontally.

![screenshot from 2015-07-07 12 53 56](https://cloud.githubusercontent.com/assets/482089/8545505/4b859898-24a7-11e5-8e24-032ef219ec5f.png)

9) Also visible in the above screenshot are odd locations for the two checkboxes, split onto multiple lines (possibly due to mark_translated?)

10) Should class and global parameters be on separate top-level tabs?

waldenraines commented on app/assets/stylesheets/hosts.scss:
```diff
@@ -35,4 +35,44 @@
+
+.push-down {
+  margin-bottom: 10px;
+}
```
I could see this establishing a bad pattern.  What happens when I need `margin-bottom: 20px;` do I create a `.push-down-2x` class?

waldenraines commented on app/views/puppetclasses/_classes_parameters.html.erb:
```diff
@@ -1,14 +1,25 @@
+$(".lookup_key_row").each(function(i, row){
+  new LookupValueController($(row));
+});
+</script>
```
Should we tab the content of this inline `<script>` tag?

unorthodoxgeek commented:
waiting for Kyle's comments for how to fix some of @domcleal's comments, pushed a fix only for the hidden flag checkbox.

domcleal commented:
See the test failures too.

unorthodoxgeek commented:
hehe, funky - all tests should have failed, for some reason the CI didn't fail on mysql and sqlite... anyway, fixed.

domcleal commented:
Integration tests are only run on PostgreSQL.

unorthodoxgeek commented:
still missing the non-refresh changes to class parameters when classes are changed. it's under development

domcleal commented:
Are all of my comments fixed, or are you still planning to change something?  I'm unsure if the flags are correct.

unorthodoxgeek commented:
@domcleal I believe that all your comments were addressed.

domcleal commented:
8) remains an issue:

![screenshot from 2015-07-14 12 35 07](https://cloud.githubusercontent.com/assets/482089/8672288/d95ff6b8-2a24-11e5-93da-50c10f24e93b.png)

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -12,7 +12,8 @@
                  :help_block => _('Whether the smart variable value is managed by Foreman')
       ) if is_param%>
-
+  <%= checkbox_f(f, :hidden, :label => _('Password field'), :size => "col-md-8",
```
I'd suggest putting this underneath the type selection.

ohadlevy commented:
we've discuessed this yesterday, I would rather remove the whole cross over
and simply provide a way to edit the original value while providing a pop
over with additional info such as original value etc.

unorthodoxgeek commented:
@domcleal - no need to review, as everything is changing once more. closing the PR for a while.

unorthodoxgeek commented:
@domcleal - can you please re-review this? thanks!

unorthodoxgeek commented:
@eLobato - maybe you can take a look as well?

dLobatog commented:
I think unfortunately it still has a long way to go. If you could post a link to the wireframes for this on the PR description to see what you have in mind that would be helpful!
Here are some of the problems I've spotted on the UI. Please feel free to dissent:

1) `Additional info` shows up on inherited parameters but not on Puppet parameters next to the information icon. I think the icon is self-explanatory enough. Think about it's placement twice too, it looks a bit misaligned next to the 'expand textbox' button. Maybe you want to add it before and after the textbox, using input-group-addon? http://getbootstrap.com/css/#optional-icons-in-horizontal-and 

![](https://www.diigo.com/file/image/rsresrqzcrbqeddqpzbrrredrb/New+Host.jpg)
![](https://www.diigo.com/file/image/rsresrqzcrbqedordzbrrredsc/New+Host.jpg)

2) When you click on it, even for very small dialogs you have to scroll. Also it's different on Global parameters, where it has a title "Additional info" that doesn't provide any information. 

![](https://www.diigo.com/file/image/rsresrqzcrbqedroazbrrreebq/New+Host.jpg)
![](https://www.diigo.com/file/image/rsresrqzcrbqeebcqzbrrreeda/New+Host.jpg)

3) Global parameters looks like a subcategory of something on the left side of the two pane, since it' uses the same CSS as Puppet params, yet it's clickable which is very misleading. I think it can simply be named "Global" (the parameters part is clear, we're in the parameters tab), and use a larger font as it's a category of it's own, not a subcategory.
After you click on it a lot of things are confusing to me at first glance:
-  The title "Included Parameters via inheritance" makes me wonder via inheritance from where? This is the menu for "Global parameters", so it means these parameters are global. It makes me wonder if I'm inheriting any of these from a host group or something. I think the title can just go away if we make a point to keep _only_ global parameters in this tab
- The scope column which says "Global" is redundant, and it takes a lot of space. We convey the same information without it if the tab says "Global" already.
- Even if I'm in the "Global parameters" section, I see something below that confuses me "Host parameters", I thought it would only be globals? Below "Host parameters" it says "Global parameters" once again, which is almost contradictory. I know what it does by heart and it still seems confusing to me. Something more clear here would be to rename it "Overrides" and only show overridden parameters. You can move Host parameters to its own section, where you show the Host specific parameters, not the overrides.

![](https://www.diigo.com/file/image/rsresrqzcrbqeoccezbrrreepr/New+Host.jpg)

4) I'm not sure two-pane is a good solution within a form. The black border around it makes it look unlike any other form we have, also unlike any other tab. It also constrains horizontal space, and makes Parameters look cramped. See `enc_mco_registration` in my example, the name won't show in full because of it. A 2nd level of tabs is what Patternfly recommends in this case. The font size of `Puppet` is for some reason bigger than anything else on the whole Host form which is confusing as I don't think it should stick out that much

![](https://www.diigo.com/file/image/rsresrqzcrbqeorcdzbrrreocs/New+Host.jpg)
![](https://www.patternfly.org/wp-content/uploads/2014/01/800px-RCUE-nav-secondary-persistent.png)

5) Overrides are the same concept for Global parameters and for Puppet parameters, yet it works in a different way on the UI. When 
- For globals, when I click on override I get a new row (fine), with no name set. This is the same behavior as 'Add Parameter' below Host Parameters. If we have a button 'override` next to every parameter, I would expect the new overridden parameter to at least take the name from the parameter next to the button I clicked on. For some reason the textbox for the original parameter and the one on the overridden one look wildly different too. As you can notice in the picture the 'full screen' button all of a sudden shows up a lot further from the textbox than in the original parameter. Also something I found confusing is that I can click the override button as many times as I want for any parameter, but it won't cross it out (it does in develop) and if I go ahead and try to override the parameter twice it won't save. I think this method of adding new rows is nicer than the one on Puppet params as you get to see what value you're overriding, The way it works on the 'old' parameters form (the current one on develop) would be great, just with some fixes for stuff like what I mentioned above.

![](https://www.diigo.com/file/image/rsresrqzcrbqeqssszbrrroadr/New+Host.jpg)
- For Puppet parameters the textbox CSS changes, and the behavior changes too as it hides the overridden value instead of adding a new row. I understand what it's going to do when I click on "override", but when it changes to "remove", it makes me thing that I'm going to completely remove that parameter, not just use the default. I'd say just use the same mechanism as in Global parameters for consistency & because it seems more intuitive.

![](https://www.diigo.com/file/image/rsresrqzcrbqereeezbrrroaoo/New+Host.jpg)
![](https://www.diigo.com/file/image/rsresrqzcrbqerepbzbrrroaoq/New+Host.jpg)

6) I don't understand why on Globals there is some text saying anything about Puppet parameters? It misses some padding on top to separate it from the Add parameter button too.

![](https://www.diigo.com/file/image/rsresrqzcrbqerepbzbrrroaoq/New+Host.jpg)

Overall it feels way more complicated, the number of levels & inconsistency in sizes, fonts, etc... I can see how this page needs a lot of help as it is bad already, thanks for taking the effort to fixing it.

orrabin commented:
I rebased this because of merge conflicts caused by my PR but @eLobato's comments aren't fixed yet. The hide button has disappeared, not sure if it has to do with the rebase but probably not since there were no  conflicts in host edit form.
Please change this back to Waiting on contributor for now.

unorthodoxgeek commented:
@orrabin - there is no checkbox, it's in the lookup key edit form. there is a bug though, as the value should be hidden and shown as a password field in the host#edit form. fixing this along with @eLobato's comments. see you next year :sob:

unorthodoxgeek commented:
@eLobato - fixed some issues, opened a new issue in redmine where out of scope
1. removed the "Additional info" text, you're right, it's irrelevant.
2. opened bug no. 11388
3. we're going to do the global parameters UI next, as this code fixes lookup keys. do you want me to do this in this PR? I suggest separating the tasks.
4. we tried the 2nd tab solution, and it was rejected
5. see 3
6. didn't understand, and screenshots are gone, can you please reup screenshots throught github, so they don't go pulling off a disappearing act? :smile: 

@orrabin - fixed the issue where hidden field did not work properly

orrabin commented:
@unorthodoxgeek I rebased this

dLobatog commented:
@unorthodoxgeek I reuploaded the pictures, sorry! Check out the comment again.

unorthodoxgeek commented:
As discussed on IRC @tbrisker will be working on this and will submit a few smaller PRs to replace this.
